extensions/amp-bind/0.1/bind-evaluator.js

```diff
@@ -121,15 +122,30 @@ export class BindEvaluator {
         errors[expressionString] = {message: error.message, stack: error.stack};
         return;
       }
+      cache[expressionString] = result;
+    });
+
+    // Then, validate each binding and delete invalid expression results.
+    this.bindings_.forEach(binding => {
+      const {tagName, property, expressionString} = binding;
+      const result = cache[expressionString];
+      if (result === undefined) {
+        return;
+      }
+      // IMPORTANT: We need to validate expression results on each binding
+      // since validity depends on the `tagName` and `property` rather than
+      // just the `result`.
       const resultString = this.stringValueOf_(property, result);
-      if (this.validator_.isResultValid(tagName, property, resultString)) {
-        cache[expressionString] = result;
-      } else {
+      if (!this.validator_.isResultValid(tagName, property, resultString)) {
+        // TODO(choumx): If this expression string is used in another
+        // tagName/property which is valid, we ought to allow it.
+        delete cache[expressionString];
```

Comment:
Yea, but the application of expression results happens in `bind-impl.js`. Parsing, evaluation and validation happen in this class, which runs in a Worker. Invalid expression results (including syntax errors, etc.) are passed back to `bind-impl.js` in an Error-like object where it's reported.

As you mention, ideally we'd only avoid applying the invalid binding (see the TODO) but this would require refactoring the `bind-impl.js` <-> Worker API. This is a fix in the interim.